src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Serialization/ObjectReader.cs

```diff
@@ -253,14 +260,31 @@ public object ReadValue()
         }
     }
 
+    public byte[] ReadByteArray()
+    {
+        var (result, _) = ReadRawArray(static length => CreateArray<byte>(length), TypeCode.UInt8, static (reader, array, length) => reader.Read(array, 0, length));
+        return result;
+    }
+
+    public char[] ReadCharArray()
+    {
+        var (result, _) = ReadCharArray(static length => CreateArray<char>(length));
+        return result;
+    }
+
     public (char[] array, int length) ReadCharArray(Func<int, char[]> getArray)
+        => ReadRawArray(getArray, TypeCode.Char, static (reader, array, length) => reader.Read(array, 0, length));
+
+    public (T[] array, int length) ReadRawArray<T>(
+        Func<int, T[]> getArray, TypeCode expectedElementKind, Func<BinaryReader, T[], int, int> read)
     {
         var kind = (TypeCode)ReadByte();
 
-        (var length, _) = ReadArrayLengthAndElementKind(kind);
+        var (length, readKind) = ReadArrayLengthAndElementKind(kind);
```

Comment:
>ReadArrayLengthAndElementKind [](http://example.com/codeflow?start=33&length=29)

Would it make sense to not put the kind in the array? Seems simpler to just read the length.

Reply:
Or maybe that's more efficient as 0-3 length entries are probably pretty common.

Reply:
yeah.  the typecode serves no purposes but to allow us to validate the reader/writer.  we don't do that elsewhere.  so i'll remove.